price-feeder/oracle/util.go

```diff
@@ -2,48 +2,115 @@ package oracle
+
+// tvwapCandlePeriod represents the time period we use for tvwap in minutes.
+const tvwapCandlePeriod = 3 * time.Minute
+
+// compute VWAP for each base by dividing the Σ {P * V} by Σ {V}
+func vwap(weightedPrices map[string]sdk.Dec, volumeSum map[string]sdk.Dec) (map[string]sdk.Dec, error) {
+	for base, p := range weightedPrices {
+		if volumeSum[base].Equal(sdk.ZeroDec()) {
+			return nil, fmt.Errorf("unable to divide by zero")
+		}
+		weightedPrices[base] = p.Quo(volumeSum[base])
+	}
+
+	return weightedPrices, nil
+}
+
 // ComputeVWAP computes the volume weighted average price for all price points
 // for each ticker/exchange pair. The provided prices argument reflects a mapping
 // of provider => {<base> => <TickerPrice>, ...}.
 //
 // Ref: https://en.wikipedia.org/wiki/Volume-weighted_average_price
 func ComputeVWAP(prices map[string]map[string]provider.TickerPrice) (map[string]sdk.Dec, error) {
 	var (
-		vwap      = make(map[string]sdk.Dec)
-		volumeSum = make(map[string]sdk.Dec)
+		weightedPrices = make(map[string]sdk.Dec)
+		volumeSum      = make(map[string]sdk.Dec)
 	)
 
 	for _, providerPrices := range prices {
 		for base, tp := range providerPrices {
-			if _, ok := vwap[base]; !ok {
-				vwap[base] = sdk.ZeroDec()
+			if _, ok := weightedPrices[base]; !ok {
+				weightedPrices[base] = sdk.ZeroDec()
 			}
 			if _, ok := volumeSum[base]; !ok {
 				volumeSum[base] = sdk.ZeroDec()
 			}
 
-			// vwap[base] = Σ {P * V} for all TickerPrice
-			vwap[base] = vwap[base].Add(tp.Price.Mul(tp.Volume))
+			// weightedPrices[base] = Σ {P * V} for all TickerPrice
+			weightedPrices[base] = weightedPrices[base].Add(tp.Price.Mul(tp.Volume))
 
 			// track total volume for each base
 			volumeSum[base] = volumeSum[base].Add(tp.Volume)
 		}
 	}
 
-	// compute VWAP for each base by dividing the Σ {P * V} by Σ {V}
-	for base, p := range vwap {
-		if volumeSum[base] == sdk.ZeroDec() {
-			return nil, fmt.Errorf("unable to divide by zero")
+	return vwap(weightedPrices, volumeSum)
+}
+
+// ComputeTVWAP computes the time volume weighted average price for all points
+// for each exchange pair. Filters out any candles that did not occur within
+// timePeriod. The provided prices argument reflects a mapping of
+// provider => {<base> => <TickerPrice>, ...}.
+//
+// Ref : https://en.wikipedia.org/wiki/Time-weighted_average_price
+func ComputeTVWAP(prices map[string]map[string][]provider.CandlePrice) (map[string]sdk.Dec, error) {
+	var (
+		weightedPrices = make(map[string]sdk.Dec)
+		volumeSum      = make(map[string]sdk.Dec)
+		now            = provider.PastUnixTime(0)
+		timePeriod     = provider.PastUnixTime(tvwapCandlePeriod)
+	)
+
+	for _, providerPrices := range prices {
+		for base, cp := range providerPrices {
+			if _, ok := weightedPrices[base]; !ok {
+				weightedPrices[base] = sdk.ZeroDec()
+			}
+			if _, ok := volumeSum[base]; !ok {
+				volumeSum[base] = sdk.ZeroDec()
+			}
+
+			// Sort by timestamp
```

Comment:
Can clarify in comment whether it's from least to recent or vice versa

Reply:
agreed 👍

Reply:
Updated to 

```
// Sort by timestamp old -> new
```